TPP.Persistence.MongoDB/Repos/PollRepo.cs

```diff
@@ -0,0 +1,132 @@
+                cm.MapProperty(b => b.PollOptions).SetElementName("options");
+                cm.MapProperty(b => b.CreatedAt).SetElementName("created_at");
+                cm.MapProperty(b => b.MultiChoice).SetElementName("multi");
+                cm.MapProperty(b => b.Alive).SetElementName("alive");
+                cm.MapProperty(b => b.AllowChangeVote).SetElementName("allow_change_vote");
+            });
+        }
+
+        public PollRepo(IMongoDatabase database, IClock clock)
+        {
+            database.CreateCollectionIfNotExists(CollectionName).Wait();
+            Collection = database.GetCollection<Poll>(CollectionName);
+            _clock = clock;
+            InitIndexes();
+        }
+
+        private void InitIndexes()
+        {
+            Collection.Indexes.CreateMany(new[]
+            {
+                new CreateIndexModel<Poll>(Builders<Poll>.IndexKeys.Ascending(u => u.PollCode)),
+                new CreateIndexModel<Poll>(Builders<Poll>.IndexKeys.Ascending(u => u.CreatedAt)),
+            });
+        }
+
+        public async Task<Poll> CreatePoll(
+            string pollCode, string pollTitle, bool multiChoice, bool allowChangeVote,
+            IImmutableList<string> pollOptions)
+        {
+            List<PollOption> pollOptionsObjects = pollOptions
+                .Select((option, index) => new PollOption(
+                    id: index + 1,
+                    option: option,
+                    voterIds: new List<string>()
+                )).ToList();
+
+            Poll poll = new(
+                pollCode: pollCode,
+                pollTitle: pollTitle,
+                voters: new List<string>(),
+                pollOptions: pollOptionsObjects,
+                _clock.GetCurrentInstant(),
+                multiChoice: multiChoice,
+                alive: true,
+                allowChangeVote: allowChangeVote
+            );
+
+            await Collection.InsertOneAsync(poll);
+
+            return poll;
+        }
+
+        public async Task<VoteFailure?> Vote(string pollCode, string userId, IImmutableList<int> options)
+        {
+            Poll? poll = await FindPoll(pollCode);
+            if (poll == null)
+                return new VoteFailure.PollNotFound(pollCode);
+            if (!poll.Alive)
+                return new VoteFailure.PollNotAlive();
+            if (!poll.AllowChangeVote && poll.Voters.Contains(userId))
+                return new VoteFailure.AlreadyVoted();
+            if (options.Count == 0)
+                return new VoteFailure.CannotVoteForNone();
+            if (options.Count > 1 && !poll.MultiChoice)
+                return new VoteFailure.NotMultipleChoice();
+
+            await Collection.UpdateOneAsync(
+                p => p.PollCode == pollCode,
+                Builders<Poll>.Update.AddToSet(p => p.Voters, userId));
+
+            ImmutableList<int> invalidOptions = options.Except(poll.PollOptions.Select(p => p.Id)).ToImmutableList();
```

Comment:
invalid options should be checked before the voter is added to the db, otherwise this breaks AllowChangeVote being set to false